hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -994,7 +994,172 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
 
   protected[ir] override def execute(ctx: ExecuteContext): TableValue = {
     val tv = child.execute(ctx)
-    agg.TableMapIRNew(ctx, tv, newRow)
+
+    val scanRef = genUID()
+    val extracted = agg.Extract.apply(CompileWithAggregators.liftScan(newRow), scanRef)
+    val nAggs = extracted.nAggs
+
+    if (extracted.aggs.isEmpty) {
+      val (rTyp, f) = ir.Compile[Long, Long, Long](
+        ctx,
+        "global", tv.globals.t,
+        "row", tv.rvd.rowPType,
+        extracted.postAggIR)
+
+      val rowIterationNeedsGlobals = Mentions(extracted.postAggIR, "global")
+      val globalsBc =
+        if (rowIterationNeedsGlobals)
+          tv.globals.broadcast
+        else
+          null
+
+      val itF = { (i: Int, ctx: RVDContext, it: Iterator[RegionValue]) =>
+        val globalRegion = ctx.freshRegion
+        val globals = if (rowIterationNeedsGlobals)
+          globalsBc.value.readRegionValue(globalRegion)
+        else
+          0
+
+        val rv2 = RegionValue()
+        val newRow = f(i, globalRegion)
+        it.map { rv =>
+          rv2.set(rv.region, newRow(rv.region, globals, false, rv.offset, false))
+          rv2
+        }
+      }
+
+      return tv.copy(
+        typ = typ,
+        rvd = tv.rvd.mapPartitionsWithIndex(RVDType(rTyp.asInstanceOf[PStruct], typ.key), itF))
+    }
+
+    val scanInitNeedsGlobals = Mentions(extracted.init, "global")
+    val scanSeqNeedsGlobals = Mentions(extracted.seqPerElt, "global")
+    val rowIterationNeedsGlobals = Mentions(extracted.postAggIR, "global")
+
+    val globalsBc =
+      if (rowIterationNeedsGlobals || scanInitNeedsGlobals || scanSeqNeedsGlobals)
+        tv.globals.broadcast
+      else
+        null
+
+    val spec = BufferSpec.defaultUncompressed
+
+    // Order of operations:
+    // 1. init op on all aggs and serialize to byte array.
+    // 2. load in init op on each partition, seq op over partition, serialize.
+    // 3. load in partition aggregations, comb op as necessary, serialize.
+    // 4. load in partStarts, calculate newRow based on those results.
+
+    val (_, initF) = ir.CompileWithAggregators2[Long, Unit](ctx,
+      extracted.aggs,
+      "global", tv.globals.t,
+      Begin(FastIndexedSeq(extracted.init, extracted.serializeSet(0, 0, spec))))
+
+    val (_, eltSeqF) = ir.CompileWithAggregators2[Long, Long, Unit](ctx,
+      extracted.aggs,
+      "global", Option(globalsBc).map(_.value.t).getOrElse(PStruct()),
+      "row", typ.rowType.physicalType,
```

Comment:
this needs to be `tv.typ` here and probably everywhere else, as well (unfortunately didn't think too hard about naming when I wrote it, since there was a `val typ = tv.typ` in the old one but here `typ` will pick up the new TableType.